lib/include/tick/array/sparsearray2d.h

```diff
@@ -106,6 +98,103 @@ class SparseArray2d : public BaseArray2d<T> {
   //! structure THUS the array becomes a view. \warning : This method cannot be
   //! called on a view
   std::shared_ptr<SSparseArray2d<T>> as_ssparsearray2d_ptr();
+
+  template <class Archive>
+  void save(Archive &ar) const {
+    try {
+      ar(this->_size_sparse);
+      ar(this->_n_rows);
+      ar(this->_n_cols);
+      ar(this->_size);
+
+      inner_save(ar);
+
+      std::vector<INDICE_TYPE> v_indices(this->_indices, this->_indices + this->_size_sparse),
+          v_row_indices(this->_row_indices, this->_row_indices + (this->_n_rows + 1));
+      ar(v_indices);
+      ar(v_row_indices);
+    } catch (const std::exception &e) {
+      std::cerr << e.what() << std::endl;
+    }
+  }
+
+  template <class Archive, typename Y = T>
+  typename std::enable_if<std::is_same<Y, std::atomic<K>>::value>::type
+  inner_save(Archive &ar) const {
+    std::vector<K> k_data;
+    for (size_t i = 0; i < this->_size_sparse; i++) k_data.emplace_back(this->_data[i]);
+    ar(k_data);
+  }
+
+  template <class Archive, typename Y = T>
+  typename std::enable_if<!std::is_same<Y, std::atomic<K>>::value>::type
+  inner_save(Archive &ar) const {
+    std::vector<T> v_data(this->_data, this->_data + this->_size_sparse);
+    ar(v_data);
+  }
+
+  template <class Archive>
+  void load(Archive &ar) {
+    if (this->_data || this->_indices || this->_row_indices)
+      throw std::runtime_error(
+          "SparseArray2d being used for deserializing may not have previous "
+          "allocations");
+
+    try {
+      ar(this->_size_sparse);
+      ar(this->_n_rows);
+      ar(this->_n_cols);
+      ar(this->_size);
+
+      // using data structures directly on SparseArray2d<T> causes segfaults
+      //  when deserializing, but using intermediary is ok, and
+      //  doesn't (seem to) cause memory leaks
+
+      T *s_data;
+      TICK_PYTHON_MALLOC(s_data, T, this->_size_sparse);
+      inner_load(ar, s_data);
+
+      INDICE_TYPE *s_indices, *s_row_indices;
+      std::vector<INDICE_TYPE> v_indices(this->_size_sparse);
+      std::vector<INDICE_TYPE> v_row_indices(this->_n_rows + 1);
+
+      TICK_PYTHON_MALLOC(s_indices, INDICE_TYPE, this->_size_sparse);
+      TICK_PYTHON_MALLOC(s_row_indices, INDICE_TYPE, this->_n_rows + 1);
+
+      ar(v_indices);
+      ar(v_row_indices);
+
+      std::copy(v_indices.begin(), v_indices.end(), s_indices);
+      std::copy(v_row_indices.begin(), v_row_indices.end(), s_row_indices);
+
+      this->_data = s_data;
+      this->_indices = s_indices;
+      this->_row_indices = s_row_indices;
+
+      this->is_data_allocation_owned = 1;
+      this->is_indices_allocation_owned = 1;
+      this->is_row_indices_allocation_owned = 1;
+    } catch (const std::exception &e) {
+      std::cerr << e.what() << std::endl;
```

Comment:
Should we catch this error? Won't this create an "unstable" object?

Reply:
we might want to test a bad file tbh - loading is more error prone than saving

Reply:
loading a bad file causes a segfault when the exception is not caught

Reply:
perhaps allowing the segfault on load is the best until we decide how to proceed